File: ai-stylist-starter/app/schemas/gap_analysis.py
"""Gap analysis response schema for GET /wardrobe/gap-analysis."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict


class GapSuggestion(BaseModel):
    """A suggested item that would unlock new outfit combinations.

    Slim UX shape:
      * ``item``   — короткое название вещи
      * ``why``    — одно короткое предложение, зачем она нужна
      * ``action`` — предлагаемое действие (кнопка)
    ``category`` сохранён для фронтенда (фильтры / иконки).

    Reference-look suggestions (Phase 7 integration) дополнительно несут
    ``from_reference_look`` (id референсного лука, под который не нашлось
    вещи в гардеробе) и ``slot_hint`` (имя слота — top/bottom/shoes/...).
    Оба поля опциональные — обычные suggestions их не выставляют.
    """

    model_config = ConfigDict(extra="forbid")

    item: str
    category: str
    why: str
    action: str = "Попробовать добавить"
    from_reference_look: str | None = None
    slot_hint: str | None = None


class UntappedItem(BaseModel):
    """An existing item that rarely appears in valid outfits."""

    model_config = ConfigDict(extra="forbid")

    item_id: str
    category: str
    outfit_count: int
    reason: str    # короткое объяснение, почему вещь редко попадает в образы


class GapAnalysisResponse(BaseModel):
    """Response for GET /wardrobe/gap-analysis."""

    model_config = ConfigDict(extra="forbid")

    suggestions: list[GapSuggestion]    # ranked by projected unlock, highest first
    untapped_items: list[UntappedItem]
    missing_categories: list[str]
    notes: list[str]


__all__ = ["GapAnalysisResponse", "GapSuggestion", "UntappedItem"]

File: frontend/src/lib/api/gapAnalysis.ts
import { z } from "zod";
import { apiRequest } from "./client";

/**
 * Backend schema is `app/schemas/gap_analysis.py::GapSuggestion`.
 * `from_reference_look` и `slot_hint` пришли из интеграции с Phase 7
 * (см. ai-stylist-starter/app/services/gap_analysis_service.py) — это
 * suggestions, рождённые из `missing_slots` референсных луков подтипа.
 */
export const GapSuggestionSchema = z.object({
  item: z.string(),
  category: z.string(),
  why: z.string(),
  action: z.string().default("Попробовать добавить"),
  from_reference_look: z.string().nullable().optional(),
  slot_hint: z.string().nullable().optional(),
});
export type GapSuggestion = z.infer<typeof GapSuggestionSchema>;

export const UntappedItemSchema = z.object({
  item_id: z.string(),
  category: z.string(),
  outfit_count: z.number(),
  reason: z.string(),
});
export type UntappedItem = z.infer<typeof UntappedItemSchema>;

export const GapAnalysisResponseSchema = z.object({
  suggestions: z.array(GapSuggestionSchema),
  untapped_items: z.array(UntappedItemSchema),
  missing_categories: z.array(z.string()),
  notes: z.array(z.string()),
});
export type GapAnalysisResponse = z.infer<typeof GapAnalysisResponseSchema>;

export async function fetchGapAnalysis(): Promise<GapAnalysisResponse> {
  const data = await apiRequest("/wardrobe/gap-analysis");
  return GapAnalysisResponseSchema.parse(data);
}
